### ISSUE TITLE
RestrictedApiChecker: java.util.NoSuchElementException at com.sun.tools.javac.code.Scope$ScopeImpl$1.next(Scope.java:592)

### ISSUE THREAD

**Issue by lgemeinhardt** [google/error-prone#2099]
### Description of the problem / feature request:

After upgrading errorprone from 2.4.0 to 2.5.0 following is reported using ant build (with JDK8):

    [javac] src\test\java\com\foo\FooTest.java:32: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
    [javac] 		final Loo helper = this.foo.new Loo() { // works if inner class is static and this.foo could be removed
    [javac] 		                            ^
    [javac]      Please report this at https://github.com/google/error-prone/issues/new and include the following:
    [javac]   
    [javac]      error-prone version: 2.5.0
    [javac]      BugPattern: RestrictedApiChecker
    [javac]      Stack Trace:
    [javac]      java.util.NoSuchElementException
    [javac]   	at com.sun.tools.javac.code.Scope$ScopeImpl$1.next(Scope.java:592)
    [javac]   	at com.sun.tools.javac.code.Scope$ScopeImpl$1.next(Scope.java:572)
    [javac]   	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:301)
    [javac]   	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
    [javac]   	at com.google.errorprone.bugpatterns.RestrictedApiChecker.superclassConstructorSymbol(RestrictedApiChecker.java:131)
    [javac]   	at com.google.errorprone.bugpatterns.RestrictedApiChecker.matchNewClass(RestrictedApiChecker.java:145)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:450)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:774)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitNewClass(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1705)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
    [javac]   	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
    [javac]   	at com.sun.source.util.TreeScanner.visitVariable(TreeScanner.java:223)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitVariable(ErrorProneScanner.java:886)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitVariable(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:968)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
    [javac]   	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
    [javac]   	at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:521)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
    [javac]   	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
    [javac]   	at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:741)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
    [javac]   	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
    [javac]   	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
    [javac]   	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
    [javac]   	at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:549)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
    [javac]   	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
    [javac]   	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
    [javac]   	at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:561)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:151)
    [javac]   	at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
    [javac]   	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
    [javac]   	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
    [javac]   	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
    [javac]   	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
    [javac]   	at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
    [javac]   	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
    [javac]   	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1353)
    [javac]   	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:946)
    [javac]   	at com.sun.tools.javac.main.Main.compile(Main.java:302)
    [javac]   	at com.sun.tools.javac.main.Main.compile(Main.java:162)
    [javac]   	at com.sun.tools.javac.Main.compile(Main.java:57)
    [javac]   	at com.sun.tools.javac.Main.main(Main.java:43)
    [javac] 1 error

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
package com.foo;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.HashSet;
import java.util.Set;

import org.junit.Test;

import com.google.common.collect.ImmutableSet;
import com.foo.FooTest.Foo.Loo;

public class FooTest {

	static class Foo {

		private boolean outer;

		class Loo { // class is non-static to show the issue

			protected Set<Integer> getGoo() {
				return Foo.this.outer ? ImmutableSet.of() : new HashSet<>(); // outer reference to avoid static
			}
		}

	}

	private final Foo foo = new Foo();

	@Test
	public void testFoo() {
		final Loo helper = this.foo.new Loo() { // works if inner class is static and this.foo could be removed

			@Override
			protected Set<Integer> getGoo() {
				return ImmutableSet.of();
			}
		};
		assertThat(this.foo).isNotNull();
		assertThat(helper).isNotNull();
	}

}
```

**This sample code is simplified and the non-static inner class raises the issue!** 

### What version of Error Prone are you using?

2.5.0

**Comment by cushon** [comment]
Thanks for the report and the repro! I can observe the same crash with this smaller example:

```java
package t;

class T {

  static class Foo {
    class Loo {}
  }

  public void testFoo(Foo foo) {
    foo.new Loo() {};
  }
}
```

```
$ javac -J-Xbootclasspath/p:javac-9+181-r4173-1.jar \
  -XDcompilePolicy=simple \
  -processorpath error_prone_core-2.5.0-with-dependencies.jar:dataflow-2.5.7.jar:javacutil-2.5.7.jar:jFormatString-3.0.0.jar \
  '-Xplugin:ErrorProne -Xep:HidingField:OFF' \
  T.java
...
T.java:10: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
    foo.new Loo() {};
        ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:

     error-prone version: 2.5.0
     BugPattern: RestrictedApiChecker
     Stack Trace:
     java.util.NoSuchElementException
  	at com.sun.tools.javac.code.Scope$ScopeImpl$1.next(Scope.java:592)
  	at com.sun.tools.javac.code.Scope$ScopeImpl$1.next(Scope.java:572)
  	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:301)  
```

**Comment by cushon** [comment]
It looks like the AST for `foo.new Loo() {};` is different in JDK 8 and 9, in 8 it might be getting desugared earlier than it should be. That is breaking this assumption:

https://github.com/google/error-prone/blob/eee9fff8423ea5525d3e2b92ca478604521de54f/core/src/main/java/com/google/errorprone/bugpatterns/RestrictedApiChecker.java#L111

**Comment by cushon** [comment]
I found the change between 8 and 9 that's causing this:

* https://github.com/openjdk/jdk/commit/b45c9233a4163edbc418a82370ed7228fe318fea
* https://bugs.openjdk.java.net/browse/JDK-8044853
